libbeat/autodiscover/providers/kubernetes/pod.go

```diff
@@ -257,194 +224,276 @@ func (p *pod) Stop() {
-				"meta": common.MapStr{
-					"kubernetes": meta,
-					"container":  cmeta,
-				},
-			}
-			events = append(events, event)
-		}
+	meta := p.metagen.Generate(pod, metadata.WithFields("container.name", c.spec.Name))
 
-		for _, port := range c.Ports {
-			podPorts[port.Name] = port.ContainerPort
-			event := bus.Event{
-				"provider":   p.uuid,
-				"id":         eventID,
-				flag:         true,
-				"host":       host,
-				"port":       port.ContainerPort,
-				"kubernetes": kubemeta,
-				"meta": common.MapStr{
-					"kubernetes": meta,
-					"container":  cmeta,
-				},
-			}
-			events = append(events, event)
-		}
-		if len(events) != 0 {
-			eventList = append(eventList, events)
-		}
+	cmeta := common.MapStr{
+		"id":      c.id,
+		"runtime": c.runtime,
+		"image": common.MapStr{
+			"name": c.spec.Image,
+		},
 	}
 
-	// Publish a pod level event so that hints that have no exposed ports can get processed.
-	// Log hints would just ignore this event as there is no ${data.container.id}
-	// Publish the pod level hint only if at least one container level hint was generated. This ensures that there is
-	// no unnecessary pod level events emitted prematurely.
-	// We publish the pod level hint first so that it doesn't override a valid container level event.
-	if len(eventList) != 0 {
-		meta := p.metagen.Generate(pod)
+	// Information that can be used in discovering a workload
+	kubemeta := meta.Clone()
+	kubemeta["annotations"] = annotations
+	kubemeta["container"] = common.MapStr{
+		"id":      c.id,
+		"name":    c.spec.Name,
+		"image":   c.spec.Image,
+		"runtime": c.runtime,
+	}
+	if len(namespaceAnnotations) != 0 {
+		kubemeta["namespace_annotations"] = namespaceAnnotations
+	}
 
-		// Information that can be used in discovering a workload
-		kubemeta := meta.Clone()
-		kubemeta["annotations"] = annotations
-		if len(nsAnn) != 0 {
-			kubemeta["namespace_annotations"] = nsAnn
-		}
+	ports := c.spec.Ports
+	if len(ports) == 0 {
+		// Ensure that at least one event is generated for this container.
+		// Set port to zero to signify that the event is from a container
+		// and not from a pod.
+		ports = []kubernetes.ContainerPort{{ContainerPort: 0}}
+	}
 
-		// Don't set a port on the event
+	var events []bus.Event
+	for _, port := range ports {
 		event := bus.Event{
 			"provider":   p.uuid,
-			"id":         fmt.Sprint(pod.GetObjectMeta().GetUID()),
+			"id":         eventID,
 			flag:         true,
-			"host":       host,
-			"ports":      podPorts,
 			"kubernetes": kubemeta,
+			// Actual metadata that will enrich the event.
 			"meta": common.MapStr{
 				"kubernetes": meta,
+				"container":  cmeta,
 			},
 		}
-		p.publish([]bus.Event{event})
+		// Include network information only if the container is running,
+		// so templates that need network don't generate a config.
+		if c.status.State.Running != nil {
+			if port.Name != "" && port.ContainerPort != 0 {
+				event["ports"] = common.MapStr{port.Name: port.ContainerPort}
```

Comment:
🤔 Will this just keep only the last "accessed" port of the loop because we create a new  `MapStr` each time?

Reply:
Yes, I have changed the behaviour here, I have moved `data.ports` from the pod event to each container/port event. Let me know if you think this can be a problem.

Before, `data.ports` was included only in the pod event, so when configurations matched they didn't include the container metadata, only the pod metadata.

Code in this PR doesn't set the whole collection of `data.ports` in the pod event, it sets a single `data.ports.<portname>` field in to each container/port event, so when configurations match they include also the container metadata. By doing this each event with a named port is going to contain a single port.

This also fixes something I found misleading during my tests. As `data.ports` was included only in pod events, templates with conditions matching container fields, like the following one, didn't work. They work now with the changes here.
```
      templates:
        - condition:
            contains.kubernetes.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.ports.redis}"
```

Reply:
You can see more clearly how this is different now in these test cases: https://github.com/elastic/beats/pull/24742/files#diff-9f24ba274864b933a726842c2b9fb8ee487a13545579938f9e373d0463611142R557

Reply:
Ok, what doesn't work now is this case. I will revert this change and do some more tests.
```
      templates:
        - condition:
            contains.kubernetes.pod.name: redis
          config:
            - module: redis
              hosts: "${data.host}:${data.ports.redis}"
            - module: nginx
              hosts: "${data.host}:${data.ports.http}"
```

We need more testing to cover all these cases :persevere:

Reply:
Ah I see, it makes sense!

Reply:
Reverted in 330182725ac779372175f3ec897451935bbe6081, I will follow with this in a separate PR.